src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs

```diff
@@ -90,94 +92,68 @@ int IVsEditorFactory.Close()
-                        var hresult = userData.SetData(
-                            VSConstants.VsTextBufferUserDataGuid.VsBufferEncodingPromptOnLoad_guid,
-                            (uint)__PROMPTONLOADFLAGS.codepagePrompt);
-
-                        if (ErrorHandler.Failed(hresult))
-                        {
-                            return hresult;
-                        }
-                    }
-                }
+                textBuffer = GetDocumentData(grfCreateDoc, pszMkDocument, vsHierarchy, itemid) as IVsTextBuffer;
+                Contract.ThrowIfNull(textBuffer, $"Failed to get document data for {pszMkDocument}");
             }
 
-            // If the text buffer is marked as read-only, ensure that the padlock icon is displayed
-            // next the new window's title and that [Read Only] is appended to title.
-            var readOnlyStatus = READONLYSTATUS.ROSTATUS_NotReadOnly;
-            if (ErrorHandler.Succeeded(textBuffer.GetStateFlags(out var textBufferFlags)) &&
-                0 != (textBufferFlags & ((uint)BUFFERSTATEFLAGS.BSF_FILESYS_READONLY | (uint)BUFFERSTATEFLAGS.BSF_USER_READONLY)))
-            {
-                readOnlyStatus = READONLYSTATUS.ROSTATUS_ReadOnly;
-            }
+            (pbstrEditorCaption, pguidCmdUI, ppunkDocView) = GetEditorData(textBuffer, physicalView, vsHierarchy, itemid, initializeView: true);
 
-            switch (physicalView)
-            {
-                case "Form":
-
-                    // We must create the WinForms designer here
-                    var loaderName = GetWinFormsLoaderName(vsHierarchy);
-                    var designerService = (IVSMDDesignerService)_oleServiceProvider.QueryService<SVSMDDesignerService>();
-                    var designerLoader = (IVSMDDesignerLoader)designerService.CreateDesignerLoader(loaderName);
-                    if (designerLoader is null)
-                    {
-                        goto case "Code";
-                    }
-
-                    try
-                    {
-                        designerLoader.Initialize(_oleServiceProvider, vsHierarchy, (int)itemid, (IVsTextLines)textBuffer);
-                        pbstrEditorCaption = designerLoader.GetEditorCaption((int)readOnlyStatus);
-
-                        var designer = designerService.CreateDesigner(_oleServiceProvider, designerLoader);
-                        ppunkDocView = Marshal.GetIUnknownForObject(designer.View);
-                        pguidCmdUI = designer.CommandGuid;
-                    }
-                    catch
-                    {
-                        designerLoader.Dispose();
-                        throw;
-                    }
+            ppunkDocData = Marshal.GetIUnknownForObject(textBuffer);
 
-                    break;
+            return VSConstants.S_OK;
+        }
+
+        public object GetDocumentData(uint grfCreate, string pszMkDocument, IVsHierarchy pHier, uint itemid)
```

Comment:
```suggestion
        public IVsTextBuffer? TryGetVsTextBuffer(uint grfCreate, string pszMkDocument, IVsHierarchy pHier, uint itemid)
```

Reply:
This is the public API. Not all EF implementations will return an IVsTextBuffer for their doc data.
GetDocumentData() and GetEditorCaption() are designed to run parts of the CreateEditorInstance() method while avoiding the unneeded parts.

Reply:
see updated comment.